dispatch.go

```diff
@@ -1,13 +1,24 @@
 package rpc
 
-import "sync"
+import (
+	"errors"
+	"sync"
+)
 
 type DecodeNext func(interface{}) error
-type ServeHook func(DecodeNext) (interface{}, error)
-type ServeNotifyHook func(DecodeNext) error
+
+type ServeHookDescription struct {
+	Args func() interface{}
```

Comment:
Maybe this function should be called `Arg` rather than `Args`?  It's one "arg" structure that contains multiple fields, so I'm not 100% convinced myself. BUT, if we call it `Args`, then the `Func` function immediately below should take a single argument called `args` rather than `arg`.